src/julia-syntax.scm

```diff
@@ -2370,7 +2370,45 @@
                 (= ,lhs ,(car rr))))
       `(= ,lhs ,rhs)))
 
-(define (expand-forms e)
+;; simple stack implementation
+(define (make-stack)
+  (let ((stack '()))
+    (lambda (op . args)
+      (case op
+        ((push) (set! stack (cons (car args) stack)))
+        ((pop)  (set! stack (cdr stack)))
+        ((top) (if (null? stack) '() (last stack)))
+        (else (error "invalid stack operation"))))))
+
+;; allocate the parsed method stack
+(define stack (make-stack))
```

Comment:
I don't think this should be using a single global stack, since that's likely to lead to difficult-to-find bugs later on. Is there a reason you can't just `cons` the method names on as `cl-convert` descends recursively?

Reply:
The issue here is the use of this stack in `expand-forms`.

As an example, consider the call-stack `expand-forms` => `expand-forms- `=> `expand-function-def` => `expand-function-def-` => `method-def-expr` => `method-def-expr`- => `current-julia-module-counter` , where A => B indicates that A calls B.

The point here is that we need to create this parsed-method-stack in `expand-forms` and use it on the transitive callee `current-julia-module-counter` .

There are two ways (I believe) of doing this:

- Using a global variable, which is what I'm doing ATM.
- Passing the parsed-method-stack as a parameter to literally all 6 transitive callees of expand-forms in this call-stack.

The second alternative can be done, but I wonder if it wouldn't be too awkward.

If we didn't have this additional use of the parsed-method-stack and the only uses were in `cl-convert` , then I'd agree that passing it as a parameter would be undoubtedly the best idea.

Reply:
As awkward as it is, the second alternative is the right way to do this IMO, because it doesn't implicitly introduce fragile global state.

As an example of the fragility I'm concerned about: If you encounter an error in femtolisp, won't this end up leaving the function-def-stack in a bad state?

Reply:
Also discussed this briefly with @NHDaly and one idea could be to use the FLISP stack itself to keep track of what's being parsed.

This has its own share of awkwardness as well, since we would essentially need to filter out the frames in the FLISP stack which are relevant for determining which is the outermost function.

Reply:
Unwinding the flisp stack dynamically does not feel like the right solution either.

Passing these arguments through manually is using the FLISP stack itself - I think it's worth the plumbing.

It's also the same as what we've done elsewhere in lowering: https://github.com/JuliaLang/julia/commit/127071cbc5d28d66a862216dc2180b29b5aa7293

Reply:
> I don't really like that this adds more hidden and secret global statefulness to the lowering algorithm as well.

Want to note that even if we do thread the `parsed-method-stack` through the callees of `expand-forms`, we won't completely get rid of the hidden state that lives inside Julia (i.e. the per-module `counter_table` hash table).